packages/next/src/server/app-render/collect-segment-data.tsx

```diff
@@ -0,0 +1,297 @@
+  serverConsumerManifest: any
+): Promise<Map<string, Buffer>> {
+  // Traverse the router tree. For each segment, decode the Flight stream for
+  // the page, pick out its segment data, and re-encode it to a new Flight
+  // stream. This will be served when performing a client-side prefetch.
+
+  // A mutable map to collect the results as we traverse the route tree.
+  const segmentBufferMap = new Map<string, Buffer>()
+  // A mutable array to collect the promises for each segment stream, so that
+  // they can run in parallel.
+  const collectedTasks: Array<Promise<void>> = []
+
+  collectSegmentDataImpl(
+    routeTree,
+    fullPageDataBuffer,
+    clientModules,
+    serverConsumerManifest,
+    [],
+    '',
+    segmentBufferMap,
+    collectedTasks
+  )
+
+  // This will resolve either after a microtask (if none of the segments
+  // have dynamic data) or in the next tick (because of the abort signal passed
+  // to renderToReadableStream).
+  await Promise.all(collectedTasks)
+
+  return segmentBufferMap
+}
+
+function collectSegmentDataImpl(
+  route: FlightRouterState,
+  fullPageDataBuffer: Buffer,
+  clientModules: ManifestNode,
+  serverConsumerManifest: any,
+  segmentPath: Array<[string, Segment]>,
+  segmentPathStr: string,
+  segmentBufferMap: Map<string, Buffer>,
+  collectedTasks: Array<Promise<void>>
+): void {
+  const children = route[1]
+  for (const parallelRouteKey in children) {
+    const childRoute = children[parallelRouteKey]
+    const childSegment = childRoute[0]
+    const childSegmentPath = segmentPath.concat([
+      [parallelRouteKey, childSegment],
+    ])
+    const childSegmentPathStr =
+      segmentPathStr +
+      '/' +
+      encodeChildSegmentAsFilesystemSafePathname(parallelRouteKey, childSegment)
+    collectSegmentDataImpl(
+      childRoute,
+      fullPageDataBuffer,
+      clientModules,
+      serverConsumerManifest,
+      childSegmentPath,
+      childSegmentPathStr,
+      segmentBufferMap,
+      collectedTasks
+    )
+  }
+
+  // Spawn a task to render the segment data to a stream.
+  collectedTasks.push(
+    renderSegmentDataToStream(
+      fullPageDataBuffer,
+      clientModules,
+      serverConsumerManifest,
+      segmentPath,
+      segmentPathStr,
+      segmentBufferMap
+    )
+  )
+}
+
+async function renderSegmentDataToStream(
+  fullPageDataBuffer: Buffer,
+  clientModules: ManifestNode,
+  serverConsumerManifest: any,
+  segmentPath: Array<[string, Segment]>,
+  segmentPathStr: string,
+  segmentBufferMap: Map<string, Buffer>
+) {
+  // Create a new Flight response that contains data only for this segment.
+  try {
+    // Since all we're doing is decoding and re-encoding a cached prerender, if
+    // it takes longer than a microtask, it must because of hanging promises
+    // caused by dynamic data. Abort the stream once the microtask queue
+    // is exhausted.
+    const abortController = new AbortController()
+    process.nextTick(() => abortController.abort())
```

Comment:
@gnoff @sebmarkbage I based this on similar code in [use-cache-wrapper when encoding the inputs](https://github.com/vercel/next.js/blob/bdf8174e7e75bc91af0e0829905c28954b726273/packages/next/src/server/use-cache/use-cache-wrapper.ts#L466-L488), and it seems work fine for the Webpack runtime. But in the Turbopack runtime it seems to abort the stream too early, always — even when there are no hanging promises.

If I change it to `setTimeout(..., 1000)` instead (or just remove the controller entirely, though obviously I need it for the cases where there actually _are_ hanging promises), then it works as expected.

I infer from this that the Turbopack is doing something async that the Webpack runtime does not. Any idea what this might be? Is there some other way I should do this that I haven't thought of?

Reply:
Fixed by warming up the module cache as in `warmFlightResponse`, per @gnoff's suggestion. Thanks!